models/dehaze_maxim/maxim.py

```diff
@@ -0,0 +1,176 @@
+
+
+import numpy as np
+import onnxruntime as ort
+from PIL import Image
+from calculate_psnr import calculate_psnr
+
+def _is_image_file(filename):
+    """Check if it is a valid image file by extension."""
+    return any(
+        filename.endswith(extension)
+        for extension in ['jpeg', 'JPEG', 'jpg', 'png', 'JPG', 'PNG', 'gif'])
+
+
+def _augment_image(image, times=8):
+    """Geometric augmentation."""
+    if times == 4:  # only rotate image
+        images = []
+        for k in range(0, 4):
+            images.append(np.rot90(image, k=k))
+        images = np.stack(images, axis=0)
+    elif times == 8:  # roate and flip image
+        images = []
+        for k in range(0, 4):
+            images.append(np.rot90(image, k=k))
+        image = np.fliplr(image)
+        for k in range(0, 4):
+            images.append(np.rot90(image, k=k))
+        images = np.stack(images, axis=0)
+    else:
+        raise Exception(f'Error times: {times}')
+    return images
+
+
+def _deaugment_image(images, times=8):
+    """Reverse the geometric augmentation."""
+
+    if times == 4:  # only rotate image
+        image = []
+        for k in range(0, 4):
+            image.append(np.rot90(images[k], k=4 - k))
+        image = np.stack(image, axis=0)
+        image = np.mean(image, axis=0)
+    elif times == 8:  # roate and flip image
+        image = []
+        for k in range(0, 4):
+            image.append(np.rot90(images[k], k=4 - k))
+        for k in range(0, 4):
+            image.append(np.fliplr(np.rot90(images[4 + k], k=4 - k)))
+        image = np.mean(image, axis=0)
+    else:
+        raise Exception(f'Error times: {times}')
+    return image
+
+
+def _save_img(img, pth):
+    """Save an image to disk.
+
+    Args:
+    img: np.ndarry, [height, width, channels], img will be clipped to [0, 1]
+        before saved to pth.
+    pth: string, path to save the image to.
+    """
+    Image.fromarray(np.array(
+        (np.clip(img, 0., 1.) * 255.).astype(np.uint8))).save(pth, 'PNG')
+
+
+def _convert_img_shape(image):
+    """Pad the image to 512*640."""
+    height, width = image.shape[0], image.shape[1]
+    padh = 512 - height
+    padw = 640 - width
+    image = np.pad(image, [(0, padh), (0, padw), (0, 0)], mode='reflect')
+    return image
+
+
+class MAXIM:
+    def __init__(self, modelPath, has_target, input_file, target_filename, geometric_ensemble, save_img, output_file):
+        self._modelPath = modelPath
+        self._create_maxim_model()
+        self._has_target = has_target
+        self._input_file = input_file
+        self._target_filename = target_filename
+        self._geometric_ensemble = geometric_ensemble
+        self._save = save_img
+        self._output_file = output_file
+
+    @property
+    def name(self):
+        return self.__class__.__name__
+
+    def _create_maxim_model(self):
+        print("Loading model....")
+        self.model = ort.InferenceSession(self._modelPath)
```

Comment:
Thanks so much for pointing this out. However, when I tried use cv2.dnn.readNetFromONNX() or cv2.dnn.readNet(). It raise an error below:

`Node [Gather@ai.onnx]:(onnx_node!jax2tf_predict_/MAXIM/jit__resize_/GatherV2_2) parse error: OpenCV(4.6.0) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:2907: error: (-215:Assertion failed) indexMat.total() == 1 in function 'parseGather'
`

Could you be so kind to give me any advice on how to solve it?

Many thanks!

Reply:
1. Simplify the model using [onnxsim](https://github.com/daquexian/onnx-simplifier), try to load using OpenCV
2. If problem persists, have a try with last OpenCV: pip install opencv-python-rolling.

If problem still persists, we will need a fix in the opencv source code.

Reply:
Thanks a lot for your very fast reply! 
- The ONNX model I tried before is already simplified. However, in case it's not, I re-run onnxsim for it. 
-  and install the last OpenCV pip install opencv-python-rolling. make sure get the successful message below:
>Successfully installed opencv-python-rolling-4.6.0.20220820

After that the same error msg still persists. In case it helps, I attached the full error msg below.

Many thanks again!

> [ERROR:0@16.068] global /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp (1044) handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Gather]:(onnx_node!jax2tf_predict_/MAXIM/jit__resize_/GatherV2_2) from domain='ai.onnx'
Traceback (most recent call last):
  File "/Users/lyn/Github/opencv_zoo/models/dehaze_maxim/demo.py", line 21, in <module>
    model = MAXIM(modelPath=args.model,
  File "/Users/lyn/Github/opencv_zoo/models/dehaze_maxim/maxim.py", line 85, in __init__
    self._create_maxim_model()
  File "/Users/lyn/Github/opencv_zoo/models/dehaze_maxim/maxim.py", line 101, in _create_maxim_model
    self.model = cv.dnn.readNetFromONNX(self._modelPath)
cv2.error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:1063: error: (-2:Unspecified error) in function 'handleNode'
> Node [Gather@ai.onnx]:(onnx_node!jax2tf_predict_/MAXIM/jit__resize_/GatherV2_2) parse error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:2645: error: (-215:Assertion failed) indexMat.total() == 1 in function 'parseGather'
>

Reply:
> parse error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:2645: error: (-215:Assertion failed) indexMat.total() == 1 in function 'parseGather'

Looks like we have a problem parsing the gather op in opencv. You can try solving it or wait a bit for my fix if possible.

Reply:
https://github.com/opencv/opencv/pull/21738 has been merged. @yunongLiu1  Could you have a try with the latest OpenCV? You could wait until Sunday as we have a weekly build releasing on each Saturday, or build latest OpenCV from source.

Reply:
Thanks a lot! And apologize for the late reply, I was a bit busy after term start. I tried to install `opencv-python-rolling-5.0.0.20220924` . Unfortunately, the same error persists.

Reply:
Updates are not forwarded to 5.x yet. So please have a try with latest 4.6: pip install “opencv-python-rolling<5.0”

Reply:
Thanks a lot for your help! I tried with 4.6. The error mentioned above was solved then. 
But, there's a new error appears, which seems like opencv does not support the op  'GreaterOrEqual' currently.

`[ERROR:0@18.681] global /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp (1019) handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [GreaterOrEqual]:(onnx_node!jax2tf_predict_/MAXIM/stage_0_encoder_block_0/channel_attention_block_10/jit_leaky_relu_/GreaterEqual) from domain='ai.onnx'
Traceback (most recent call last):
  File "demo.py", line 21, in <module>
    model = MAXIM(modelPath=args.model,
  File "/Users/lyn/Library/Mobile Documents/com~apple~CloudDocs/.Trash/opencv_zoo 3/models/dehaze_maxim/maxim.py", line 85, in __init__
    self._create_maxim_model()
  File "/Users/lyn/Library/Mobile Documents/com~apple~CloudDocs/.Trash/opencv_zoo 3/models/dehaze_maxim/maxim.py", line 101, in _create_maxim_model
    self.model = cv.dnn.readNetFromONNX(self._modelPath)
cv2.error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:1038: error: (-2:Unspecified error) in function 'handleNode'
> Node [GreaterOrEqual@ai.onnx]:(onnx_node!jax2tf_predict_/MAXIM/stage_0_encoder_block_0/channel_attention_block_10/jit_leaky_relu_/GreaterEqual) parse error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/layer_internals.hpp:110: error: (-2:Unspecified error) Can't create layer "onnx_node!jax2tf_predict_/MAXIM/stage_0_encoder_block_0/channel_attention_block_10/jit_leaky_relu_/GreaterEqual" of type "GreaterOrEqual" in function 'getLayerInstance'
>`

Reply:
https://github.com/opencv/opencv/pull/22611 for supporting GreaterOrEqual has been merged. Please have a try after this Saturday.

Reply:
Thanks for reminder :) The new version did solve the previous problem. I encountered another problem then.

> [ERROR:0@16.696] global /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp (1050) handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Gemm]:(onnx_node!Gemm__9265) from domain='ai.onnx'
Traceback (most recent call last):
  File "demo.py", line 21, in <module>
    model = MAXIM(modelPath=args.model,
  File "/Users/lyn/Documents/opencv_zoo/models/dehaze_maxim/maxim.py", line 85, in __init__
    self._create_maxim_model()
  File "/Users/lyn/Documents/opencv_zoo/models/dehaze_maxim/maxim.py", line 101, in _create_maxim_model
    self.model = cv.dnn.readNetFromONNX(self._modelPath)
cv2.error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:1069: error: (-2:Unspecified error) in function 'handleNode'
> Node [Gemm@ai.onnx]:(onnx_node!Gemm__9265) parse error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:587: error: (-5:Bad argument) Blob Reshape__9264:0 not found in const blobs in function 'getBlob'
> 
(pythonProject2) lyn@pc-66-62 dehaze_maxim % python demo.py
Loading model....
[ERROR:0@15.250] global /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp (1050) handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Gemm]:(onnx_node!Gemm__9265) from domain='ai.onnx'
Traceback (most recent call last):
  File "demo.py", line 21, in <module>
    model = MAXIM(modelPath=args.model,
  File "/Users/lyn/Documents/opencv_zoo_20221022/models/dehaze_maxim/maxim.py", line 85, in __init__
    self._create_maxim_model()
  File "/Users/lyn/Documents/opencv_zoo_20221022/models/dehaze_maxim/maxim.py", line 101, in _create_maxim_model
    self.model = cv.dnn.readNetFromONNX(self._modelPath)
cv2.error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:1069: error: (-2:Unspecified error) in function 'handleNode'
> Node [Gemm@ai.onnx]:(onnx_node!Gemm__9265) parse error: OpenCV(4.6.0-dev) /Users/runner/work/opencv-python/opencv-python/opencv/modules/dnn/src/onnx/onnx_importer.cpp:587: error: (-5:Bad argument) Blob Reshape__9264:0 not found in const blobs in function 'getBlob'

i left it until today since it seems not like a problem in the source code and I was distracted by my university courses.  Apologize for the delay. Do you have any suggestions at this stage? Many thanks